ciao-controller/cnci.go

```diff
@@ -0,0 +1,579 @@
+	cnci.instance = nil
+
+	return nil
+}
+
+func (c *CNCIManager) waitForActive(subnet int) error {
+	c.cnciLock.RLock()
+
+	cnci, ok := c.subnets[subnet]
+
+	c.cnciLock.RUnlock()
+
+	if !ok {
+		return errors.New("No CNCI found")
+	}
+
+	if instanceActive(cnci.instance) {
+		return nil
+	}
+
+	// lock eventCh
+	cnci.eventChLock.Lock()
+	eCh := cnci.eventCh
+	cnci.eventChLock.Unlock()
+
+	// CNCI launch not in process, and it's not active.
+	if eCh == nil {
+		return errors.New("CNCI not active")
+	}
+
+	// CNCI launch in process. we wait here till
+	// the channel is closed. When it is, the cnci
+	// is either active, or it failed to start.
+	<-*eCh
+	if instanceActive(cnci.instance) {
+		return nil
+	}
+
+	return errors.New("CNCI not active")
+}
+
+// GetInstanceCNCI will return the CNCI Instance for a specific tenant Instance
+func (c *CNCIManager) GetInstanceCNCI(ID string) (*types.Instance, error) {
+	// figure out what subnet we are looking for.
+	instance, err := c.ctrl.ds.GetInstance(ID)
+	if err != nil {
+		return nil, err
+	}
+
+	// convert subnet string to int
+	subnetInt, err := subnetStringToInt(instance.Subnet)
+	if err != nil {
+		return nil, err
+	}
+
+	c.cnciLock.Lock()
+	defer c.cnciLock.Unlock()
+
+	cnci, ok := c.subnets[int(subnetInt)]
+	if !ok {
+		// there is no cnci for this subnet
+		return nil, errors.New("Subnet doesn't exist")
+	}
+
+	return cnci.instance, nil
+}
+
+func subnetStringToInt(cidr string) (int, error) {
+	ipAddr, _, err := net.ParseCIDR(cidr)
+	if err != nil {
+		return 0, err
+	}
+
+	ipBytes := ipAddr.To4()
+	if ipBytes == nil {
+		return 0, errors.New("Unable to convert ip to bytes")
+	}
+
+	return int(binary.BigEndian.Uint16(ipBytes[1:3])), nil
+}
+
+// GetSubnetCNCI will return the CNCI Instance for a specific subnet string
+func (c *CNCIManager) GetSubnetCNCI(subnet string) (*types.Instance, error) {
+	subnetInt, err := subnetStringToInt(subnet)
+	if err != nil {
+		return nil, err
+	}
+
+	c.cnciLock.Lock()
+	defer c.cnciLock.Unlock()
+
+	cnci, ok := c.subnets[subnetInt]
+	if !ok {
+		// there is no cnci for this subnet
+		return nil, errors.New("Subnet doesn't exist")
+	}
+
+	return cnci.instance, nil
+}
+
+func (c *CNCIManager) getInstanceCount(subnet string) (int, error) {
+	var count int
+
+	instances, err := c.ctrl.ds.GetAllInstancesFromTenant(c.tenant)
+	if err != nil {
+		return 0, err
+	}
+
+	for _, i := range instances {
+		if i.Subnet == subnet {
+			count++
+		}
+	}
+
+	return count, nil
+}
+
+// Shutdown cleans up a CNCIManager in anticipation of a shutdown.
+func (c *CNCIManager) Shutdown() {
+	// the only thing we need to do right now at shutdown time
+	// is to make sure any in progress timers are cancelled.
+	for _, cnci := range c.subnets {
+		if cnci.timer != nil {
+			cnci.timer.Stop()
+			cnci.timer = nil
+		}
+	}
+}
+
+func newCNCIManager(ctrl *controller, tenant string) (*CNCIManager, error) {
+	mgr := CNCIManager{
+		tenant: tenant,
+		ctrl:   ctrl,
+
+		cnciLock: &sync.RWMutex{},
```

Comment:
Is there a reason to use a pointer here?

Reply:
no.